spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -511,7 +511,13 @@ trait OptimisticTransactionImpl extends TransactionalWrite
 
       val newProtocolForLatestMetadata =
         Protocol(readerVersionAsTableProp, writerVersionAsTableProp)
-      val proposedNewProtocol = protocolBeforeUpdate.merge(newProtocolForLatestMetadata)
+      // When the protocol versions are explicitly set on table features protocol we may
+      // downgrade to legacy protocol versions.
+      val proposedNewProtocol = protocolBeforeUpdate
+        .merge(newProtocolForLatestMetadata)
+        .downgradeProtocolToLegacyVersionsIfNeeded(
```

Comment:
OK, so in this case we only downgrade to the explicitly specified protocol, NOT to _any_ legacy protocol. E.g. if the table contains the feature added by legacy protocol 6, but the caller asks to go down to protocol 5, then we won't downgrade -- you have to _exactly_ hit the target protocol version. Should we instead just go all the way and do a a downgrade to _any_ legacy version when people ask for a legacy version?

Reply:
Done.